lib/change_stream.js

```diff
@@ -561,8 +539,65 @@ function processNewChange(args) {
 
   // Return the change
   if (eventEmitter) return changeStream.emit('change', change);
-  if (typeof callback === 'function') return callback(error, change);
-  return changeStream.promiseLibrary.resolve(change);
+  return callback(error, change);
+}
+
+/**
+ * Safely provides a cursor across resume attempts
+ *
+ * @param {ChangeStream} changeStream the parent ChangeStream
+ * @param {function} callback gets the cursor or error
+ * @param {ChangeStreamCursor} [oldCursor] when resuming from an error, carry over options from previous cursor
+ */
+function getCursor(changeStream, callback, oldCursor) {
+  // if a cursor exists and it is open, return it
+  if (changeStream.cursor) {
+    if (changeStream.isClosed()) {
+      callback(new MongoError('Change Stream is not open.'));
+      return;
+    }
+    callback(undefined, changeStream.cursor);
+    return;
+  }
+
+  // no cursor, queue callback until topology reconnects
+  changeStream[kResumeQueue].push(callback);
+  if (
+    changeStream.waitingForTopology || // don't reconnect if already reconnecting
+    !oldCursor // only reconnect if the previous cursor was provided
+  ) {
+    return;
+  }
+
+  // attempt to reconnect the topology
```

Comment:
maybe we can refactor this method into `makeChangeStreamCursor` and call it from `getCursor` if no cursor is available, as well as from `processNewChange` in the error case? I think this would make it so we don't have to pass `oldCursor` into `getCursor`

Reply:
Not exactly what you described, but I extracted a new `recreateCursor`, which is called inside `processNewChange`. I also renamed `createChangeStreamCursor` to `createCursor` for brevity/consistency with `getCursor` and `recreateCursor`.